Make Options::apply fallible

BenWiederhake opened the pull request:
This PR shows one way we could make `Options::apply` fallible.

This feature trivially enables error prioritization, e.g. raising different errors for `--a=invalid --b=invalid` and `--b=invalid --a=invalid`, which is what the GNU tools do.

Closes #112

BenWiederhake commented:
Changes since initial push:
- Add `date` example, since this usecase motivated this change.
- Fix a test in `shuf` that exposes the same problem I noticed while writing `date` test: `--help` exits the entire test runner, thus hiding potential test failures. This is a workaround that needs a proper fix later.
- Something with clippy changed, it now complains about code I didn't even touch. So I fixed it by adding `#[allow(unused)]` to a handful of things.

BenWiederhake commented:
Cahnges since last push: `cargo fmt` :sweat_smile:

BenWiederhake commented:
Changes since last push:
- Rebased
- Resolved trivial conflict (both this PR and #122 added a test module)

BenWiederhake commented:
Changes since last push:
- Resolved another trivial conflict (forgot to add the return-type in `Settings::apply` for the cksum-test)
- Fixed the date value-prefix logic. (We should make some nicer logic than this, but it's out-of-scope for this PR.)

tertsdiepraam commented on tests/coreutils/date.rs:
```diff
@@ -0,0 +1,712 @@
+enum Iso8601Format {
+    #[default]
+    #[value("date")]
+    // TODO: Express the concept "accepts prefixes" more nicely.
```
I thought I built this into `Value` already? It's working here for example:

https://github.com/uutils/uutils-args/blob/a3ecc624a93ebd592a885567162f8cebb8fa4178/tests/coreutils/ls.rs#L480

BenWiederhake replied:
You're right! Fixed

tertsdiepraam commented on tests/coreutils/date.rs:
```diff
@@ -0,0 +1,712 @@
+
+impl Options<Arg> for Settings {
+    fn apply(&mut self, arg: Arg) -> Result<(), uutils_args::Error> {
+        if self.chosen_format != Format::Unspecified {
```
So, I take a bit of issue with this example, because I don't like how `Unspecified` is a special thing that only exists for parsing. Essentially, it comes down to this: https://github.com/uutils/uutils-args/issues/11.

BUT! I don't care enough to block this. We can improve this later and we should move towards shipping this crate.

tertsdiepraam reviewed (changes requested):
Awesome! My only real issue is with the `Value` thing I commented on. Once that's resolved: LGTM!

BenWiederhake commented:
Changes since last push:
- Removed the superfluous manual alias-definitions

The clippy failure will be fixed by #125. Sorry, didn't see that it's my turn!
